### PR TITLE
Faster case-insensitive comparisons.

### PR DESCRIPTION
**Pull request by lemire**
Ok. So this thing comes and goes. I think I need to convince @jkeiser that "my way" is better.

I wrote a whole blog post...

[For case-insensitive string comparisons, avoid char-by-char functions](https://lemire.me/blog/2020/04/30/for-case-insensitive-string-comparisons-avoid-char-by-char-functions/)

There is a benchmark and everything.

The tolower approach is a dog: 4ns/byte (on two systems). That's > 10 cycles to compare two bytes.

It is also more lines of code.

It is bad, bad, bad.

Am I convincing enough?

:smile:

### CODE REVIEW

**Comment by jkeiser** [comment]
To be clear, I was never really confident my way would be good, I just had a ton of other code to write for DOM API and didn't spend a lot of time on it :) I really appreciate the deep look!

File: include/simdjson/document.h [hunk 1]
```diff
@@ -313,6 +313,7 @@ class object : protected internal::tape_ref {
 
   /**
    * Get the value associated with the given key in a case-insensitive manner.
+   * It is only guaranteed to work over ASCII inputs.
```
Comment (jkeiser) [inline include/simdjson/document.h:316]: This is a fascinating, and thoroughly reasonable, place to land :)

**Comment by jkeiser** [comment]
Also, why the #!@#$!$ are we running checkperf on sanitize builds? I thought I had that locked down. I'll look at it.

@lemire you are green on CI, the failure is fake.

**Comment by lemire** [comment]
@jkeiser 


My wife would no doubt insist that I explain.

So I was trying to be funny. I realize that you know to expect that the single function will be faster. :-)

**Comment by jkeiser** [comment]
lol I think the fact that you wrote a whole blog post threw me off the sarcasm :)

**Comment by lemire** [comment]
> lol I think the fact that you wrote a whole blog post threw me off the sarcasm :)


I realize this. But if you go on twitter, you will see that there are disagreements as to whether the char-by-char approach is really inferior. So it is not an obvious fact. I think that libc++ does a character-by-character approach insider. Only glibc does something different.

I have spent hours trying to beat glibc without using SIMD, and I have not managed to do it.

**Comment by lemire** [comment]
So you could have said "I think it is just as fast", and you would have a point... possibly... I couldn't know before I benchmarked it.

**Pull request merged**